river/tree/splitter/sgt_quantizer.py

```diff
@@ -0,0 +1,182 @@
+
+    Parameters
+    ----------
+    n_bins
+        The number of bins (intervals) to divide the input feature.
+    warm_start
+        The number of observations used to initialize the quantization intervals.
+    buckets
+        This parameter is only used internally by the quantizer, so it must not be set.
+        Once the intervals are defined, new instances of this quantizer will receive the
+        quantization information via this parameter.
+
+    References
+    ----------
+    [^1]: Gouk, H., Pfahringer, B., & Frank, E. (2019, October). Stochastic Gradient Trees.
+    In Asian Conference on Machine Learning (pp. 1094-1109).
+
+    """
+
+    def __init__(
+        self, n_bins: int = 64, warm_start: int = 100, *, buckets: typing.List = None
+    ):
+        super().__init__()
+
+        self.n_bins = n_bins
+        self.warm_start = warm_start
+        self.buckets = buckets
+
+        if self.buckets is None:
+            self._buffer = []
+            self._min = None
+            self._radius = None
+        else:
+            self._buffer = None
+            # Define the quantization radius and the minimum value to data perform shift
+            self._radius = self.buckets[1][0][1] - self.buckets[1][0][0]
+            self._min = self.buckets[0][0][1] - self._radius
+
+    def update(self, x_val, gh: GradHess, w: float):
+        if self.buckets is None:
+            self._buffer.append((x_val, gh, w))
+
+            if len(self._buffer) < self.warm_start:
+                return
+
+            self._min = min(self._buffer, key=lambda t: t[0])[0]
+            _max = max(self._buffer, key=lambda t: t[0])[0]
+            self._radius = (_max - self._min) / self.n_bins
+
+            splits = (
+                [-math.inf]
+                + [self._min + i * self._radius for i in range(1, self.n_bins)]
+                + [math.inf]
+            )
+
+            self.buckets = [
+                ((splits[i], splits[i + 1]), GradHessStats())
+                for i in range(self.n_bins)
+            ]
+
+            # Replay buffer
+            for x_val, gh, w in self._buffer:
+                pos = math.floor((x_val - self._min) / self._radius)
+                if pos >= self.n_bins:
+                    pos = self.n_bins - 1
+                self.buckets[pos][1].update(gh, w)
+
+            # And empty it
+            self._buffer = None
+
+        # Projection scheme
+        pos = math.floor((x_val - self._min) / self._radius)
+        if pos < 0:
+            pos = 0
+        if pos >= self.n_bins:
+            pos = self.n_bins - 1
+
+        # Update the corresponding bucket
+        self.buckets[pos][1].update(gh, w)
+
+    def __len__(self):
+        if self.buckets:
+            return len(self.buckets)
+        return 0
+
+    def __iter__(self):
+        # TODO check that: in this case the null split should be choice
+        if self.buckets is None:
+            return
```

Comment:
To verify but I think you need to add a `yield` before or after the `return` to make this method a generator.

Reply:
Hmm, thanks for the tip! I'll check that :D